aws/logs_monitoring/lambda_function.py

```diff
@@ -652,7 +675,35 @@ def awslogs_handler(event, context, metadata):
                 metadata[DD_HOST] = arn
 
     # Create and send structured logs to Datadog
-    for log in logs["logEvents"]:
+    # if multiline regex, iterate over log events and concatenate multiline logs into parent log
+    # child multiline logs must have matching timestamp and not match regex start pattern
+    log_events = []
+
+    if DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS and logs["logGroup"] in DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS:
+        multiline_log_timestamp = None
+        multiline_log_flag = False
+        log_group = logs["logGroup"]
+
+        for log in logs["logEvents"]:
+            if DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS[log_group].match(log["message"]):
+                multiline_log_timestamp = log["timestamp"]
+                multiline_log_flag = True
+                log_events.append(log)
+            elif log["timestamp"] != multiline_log_timestamp:
+                multiline_log_timestamp = None
+                multiline_log_flag = False
+                log_events.append(log)
+            elif multiline_log_flag:
+                # valid child multiline log, matching timestamp and multiline log flag has been toggled
+                log_events[-1]["message"] += log["message"]
```

Comment:
could we use an intermediate buffer instead of appending the content directly to the last log please ?

Reply:
👍  agree, I updated this to use an empty dictionary `{}` as intermediate storage here of the multiline log, which let me clean up `if/elif/else` code here.  Apologies as I wasn't completely sure what you meant by buffer so if there's a better or more pythonic way of doing this please lmk, can update